modules/core/05-port/types/ibc_legacy_module.go

```diff
@@ -435,17 +434,32 @@ func (im *LegacyIBCModule) OnChanUpgradeOpen(ctx sdk.Context, portID, channelID
 // UnmarshalPacketData attempts to unmarshal the provided packet data bytes
 // into a FungibleTokenPacketData. This function implements the optional
 // PacketDataUnmarshaler interface required for ADR 008 support.
-func (LegacyIBCModule) UnmarshalPacketData(ctx sdk.Context, portID, channelID string, bz []byte) (interface{}, error) {
+func (*LegacyIBCModule) UnmarshalPacketData(ctx sdk.Context, portID, channelID string, bz []byte) (interface{}, error) {
 	return nil, nil
 }
 
+// reverseCallbacks returns a copy of the callbacks in reverse order.
+// the majority of handlers are called in reverse order, so this can be used
+// in those cases to prevent needing to iterate backwards over the callbacks.
+func (im *LegacyIBCModule) reverseCallbacks() []ClassicIBCModule {
+	cbs := slices.Clone(im.cbs)
+	slices.Reverse(cbs)
+	return cbs
+}
+
 // reconstructVersion will generate the channel version by applying any version wrapping as necessary.
 // Version wrapping will only occur if the negotiated version is non=empty and the application is a VersionWrapper.
-func reconstructVersion(cbs []ClassicIBCModule, negotiatedVersions []string) (string, error) {
+func (im *LegacyIBCModule) reconstructVersion(negotiatedVersions []string) (string, error) {
+	// the negotiated versions are expected to be in reverse order, as callbacks are executed in reverse order.
+	// in order to ensure that the indices match im.cbs, they must be reversed.
+	// the slice is cloned to prevent modifying the input argument.
+	negotiatedVersions = slices.Clone(negotiatedVersions)
+	slices.Reverse(negotiatedVersions)
```

Comment:
this is the one thing I don't like about this PR, the negotiatedVersions are provided in reverse order.

I think we can either do this, require the caller to provide the calling fn to reverse them (easy to forget), or just reverse directly without copying, which will likely not be a problem but this version is just being extra careful.

Reply:
Hmm, yeah. Not beautiful, but its all part of this legacy module construction so I'm fine with it